### ISSUE TITLE
ORM, Double to string conversion and internationalization (Very serious).

### ISSUE THREAD

**Issue by Josua2012** [silverstripe/silverstripe-framework#2606]
I use ecommerce module and have noticed a problem that seems to be due to framework.

In my spanish server, when I select another currency the following function is execute and error occurs:

```
    public function UpdateCurrency($newCurrency) {
        if($this->IsSubmitted()) {
            user_error("Can not set the exchange rate after the order has been submitted", E_USER_NOTICE);
        }
        else {
            if(!is_a($newCurrency, Object::getCustomClass("EcommerceCurrency"))) {
                $currency = EcommerceCurrency::default_currency();
            }
            $this->CurrencyUsedID = $newCurrency->ID;
            $this->ExchangeRate = $newCurrency->ExchangeRate();
            $this->write();
        }
    }
```

[User Error] Couldn't run query: UPDATE "Order" SET "CurrencyUsedID" = 3, "ExchangeRate" = 1,44942, "LastEdited" = '2013-10-24 20:10:21' where "ID" = 903 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '44942, "LastEdited" = '2013-10-24 20:10:21' where "ID" = 903' at line 1
GET /shoppingcart/setcurrency/USD/?SecurityID=3b0a38496ab63fde877047306a1504fba1577b1d

Line 592 in /var/www/vhosts/sendasoft.com/httpdocs/framework/model/MySQLDatabase.php
Source

583     }
584 
585     public function databaseError($msg, $errorLevel = E_USER_ERROR) {
586         // try to extract and format query
587         if(preg_match('/Couldn\'t run query: ([^|]_)|\s_(._)/', $msg, $matches)) {
588             $formatter = new SQLFormatter();
589             $msg = "Couldn't run query: \n" . $formatter->formatPlain($matches[1]) . "\n\n" . $matches[2];
590         }
591 
592         user_error($msg, $errorLevel);
593     }
594 
595     /_*
596      \* Return a boolean type-formatted string
597      *
598      \* @param array $values Contains a tokenised list of info about this data type

Trace

```
Couldn't run query: UPDATE "Order" SET "CurrencyUsedID" = 3, "ExchangeRate" = 1,44942, "LastEdited" = '2013-10-24 20:10:21' where "ID" = 903 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '44942, "LastEdited" = '2013-10-24 20:10:21' where "ID" = 903' at line 1
MySQLDatabase.php:592
MySQLDatabase->databaseError(Couldn't run query: UPDATE "Order" SET "CurrencyUsedID" = 3, "ExchangeRate" = 1,44942, "LastEdited" = '2013-10-24 20:10:21' where "ID" = 903 | You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '44942, "LastEdited" = '2013-10-24 20:10:21' where "ID" = 903' at line 1,256)
MySQLDatabase.php:144
MySQLDatabase->query(UPDATE "Order" SET "CurrencyUsedID" = 3, "ExchangeRate" = 1,44942, "LastEdited" = '2013-10-24 20:10:21' where "ID" = 903)
Database.php:633
SS_Database->manipulate(Array)
DB.php:245
DB::manipulate(Array) 
```

The problem occurs for the conversion of the ExchangeRate amount to string:
"ExchangeRate" = 1,44942

This should be:
"ExchangeRate" = 1.44942

Regards,
Jose A.

**Comment by chillu** [comment]
- We've rewritten core to use [prepared statements](https://github.com/silverstripe/silverstripe-framework/issues/1486), meaning there's no more use of string casting through `real_escape_string()`
- `DBFloat->prepValueForDB()` uses `is_numeric()`, which isn't dependant on `setlocale()`
- `number_format()` is used around core a bit, but again isn't dependant on `setlocale()`
- `strftime()` is dependant on `setlocale()`, but we don't use it in core (only in some tests)

I've tried writing setting `setlocale(LC_ALL, 'de')`, which uses commas instead of dots for decimal separators. Wrote a `DBFloat` value into a database record (`$myRecord->MyFloat = 12.34; $myRecord->write()`), works as expected. If you're specifically passing a string with the wrong decimal separator into there (`"12,34"`), `DBFloat` will simply write `0` - which is expected behavior. You shouldn't rely on implicit casting there. If you're using `NumericField` through forms, just define which locale you're accepting, and make the conversion explicit. 

So in short, I think this is solved simply by not relying on PHP's string-to-float casting in core any longer.
